File: Learning/src/DataStructures/QueueExample.java
package DataStructures;
import java.util.LinkedList;
import java.util.Queue;

public class QueueExample {
    public static void main(String[] args) {
        // Signature
        Queue<Integer> queue = new LinkedList<>();

        // Enqueue
        System.out.println("Enqueuing element: 1");
        queue.add(1);

        // Dequeue
        if (!queue.isEmpty()) {
            int removedElement = queue.poll();
            System.out.println("Dequeued element: " + removedElement);
        } else {
            System.out.println("Queue is empty. Cannot dequeue.");
        }

        // Peek
        if (!queue.isEmpty()) {
            int frontElement = queue.peek();
            System.out.println("Front element: " + frontElement);
        } else {
            System.out.println("Queue is empty. No front element to peek.");
        }

        // Add more elements
        queue.add(2);
        queue.add(3);


        // Length (Size)
        System.out.println("Size of Queue: " + queue.size());

        // Iterate through Queue using for-each loop
        System.out.println("\nIterating through Queue using for-each loop:");
        for (int element : queue) {
            System.out.println("Element: " + element);
        }

        // Alternatively, use Iterator for more control
        System.out.println("\nIterating using Iterator:");
        java.util.Iterator<Integer> iterator = queue.iterator();
        while (iterator.hasNext()) {
            int element = iterator.next();
            System.out.println("Element: " + element);
        }

        // Alternatively, use forEach method (Java 8 and later)
        System.out.println("\nIterating using forEach method:");
        queue.forEach(element -> System.out.println("Element: " + element));

        // Alternatively, convert to array and use enhanced for loop
        System.out.println("\nIterating using toArray method:");
        Integer[] elementsArray = queue.toArray(new Integer[0]);
        for (int element : elementsArray) {
            System.out.println("Element: " + element);
        }

    }
}

File: DSA/Python/DataStructures/queue_example.py
from collections import deque

def main():
    # Initialize queue
    queue = deque()

    # Enqueue
    print("Enqueuing element: 1")
    queue.append(1)

    # Dequeue
    if queue:
        removed_element = queue.popleft()
        print(f"Dequeued element: {removed_element}")
    else:
        print("Queue is empty. Cannot dequeue.")

    # Peek
    if queue:
        front_element = queue[0]
        print(f"Front element: {front_element}")
    else:
        print("Queue is empty. No front element to peek.")

    # Add more elements
    queue.append(2)
    queue.append(3)
    print(f"Is the queue empty? {len(queue) == 0}")

    # Length (Size)
    print(f"Size of Queue: {len(queue)}")

    # Iterate through Queue using for loop
    print("\nIterating through Queue using for loop:")
    for element in queue:
        print(f"Element: {element}")

    # Iterate using enumerate
    print("\nIterating using enumerate:")
    for i, element in enumerate(queue):
        print(f"Element: {element}")

    # Iterate using list comprehension
    print("\nIterating using list comprehension:")
    [print(f"Element: {element}") for element in queue]

    # Convert to list and iterate
    print("\nIterating using converted list:")
    elements_list = list(queue)
    for element in elements_list:
        print(f"Element: {element}")

if __name__ == "__main__":
    main()
